### ISSUE TITLE
Improve signal items

### ISSUE THREAD

**Issue by npiganeau** [ts2/ts2#15]
- Signal items should have an almost null length on the layout
=> Need to separate signals and berths 
- Add other signal types, like 4-aspect signals, shunt signals, etc.

**Comment by npiganeau** [comment]
Just for info, I'm currently working on the signal system where there will be a "signal library" loaded from JSON files to be used by all simulations. The idea is that anyone could create custom signals (for diferent countries for instance) just by creating new JSON files.

I'm also thinking of being able to declare signals directly into the simulation too so that we can share the simulation without needing any dependency, but that's much more complex. I think I will be able to manage it for the game, but not for the editor, which would mean that for a simulation with custom signals:
- You could play it just by downloading the simulation file
- You would need to download the signal library file too to be able to edit it.

Or we could create a system of dependency from ts2-data...

**Comment by npiganeau** [comment]
More news on the new signal configuration of TS2 as this is now merged into the develop branch.

### What is it all about ?

Signal types (such as UK_3_ASPECTS, UK_4_ASPECTS but even BUFFER) are now defined in JSON configuration files and are not hardcoded anymore. This allows anybody to create its custom signals and the way these signals will behave in the simulation without the need to hack into TS2. It also allows to create such custom signals very quickly. 

### How does it work?

New signal types are defined in .tsl files in the data directory. All .tsl files are loaded when the application is started and are merged to create the signal library holding all available signal types.

TSL files are just plain JSON. There format is:
```JSON
{
    "__type__": "SignalLibrary",
    "signalAspects": {
...
    },
    "signalTypes": {
...
    }
}
```
This defines two type of objects:
- signalAspects is a library of aspects that signals can show. The same aspect can be used by different Signal Types in the same file or in other TSL files.
- signalTypes which defines all the different available types for a signal Item when editing it in a simulation file. A signal type defines the different aspects that a signal can show with the conditions for them to show.

#### Signal aspects
Signal aspects in TS2 have 6 possible "light bulbs" (three rows of 2) which are numbered:
- 1 at the bottom right next to the signal pole
- 2 at the left of bulb 1
- 3 above bulb 1
- 4 above bulb 2
- 5 above bulb 3
- 6 above bulb 4

The ```signalAspects``` key takes a dictionary as value whose:
- Keys are ```aspect codes```
- Values are ```aspect definition``` dictionary

```aspect definition``` takes 7 key-value pairs:
- ```__type__``` must be equal to "SignalAspect"
- ```lineStyle``` is 1 for buffers, 0 for all other signals
- ```outerShapes``` is a list of 6 integers defining the outer shape (background shape) of each of the 6 bulbs. Signification of each integer is given here: https://github.com/ts2/ts2/blob/develop/ts2/scenery/signals/signalaspect.py#L24
- ```outerColors``` is a list of 6 colors to fill the outer shape of each bulb. #RRGGBB format
- ```shapes``` is a list of 6 integers to draw the actual shape of each bulb. Same codes as ```outerShapes```
- ```shapesColors``` is a list of 6 colors to fill the shape of each bulb. #RRGGBB format 
- ```actions``` is a list of actions that the train must perform in turn when meeting this aspect (For now only the first action is executed). Each action is actually conforming to a speed limit (which can be 0) and is defined by a list of two numbers:
  - The first one is the target at which the action (speed limit) must be performed. It can be 0: ASAP, 1: Before this signal, 2: before next signal
  - The second one is the speed in metres / second to achieve at the target

#### Signal Types
A signal type is defined by its possible states. A state is an aspect together with the conditions for this aspect to show.

The ```signalTypes``` key takes a dictionary as value whose:
- Keys are ```type codes```
- Values are ```type definition``` dictionary

```type definition``` dictionary takes only 2 keys:
- ```__type__``` must be equal to "SignalType"
- ```states``` which is a list of ```state``` dictionaries

```state``` dictionary takes 3 keys:
- ```__type__``` which must be equal to "SignalState"
- ```aspectName``` is a string which must match one of the keys of ```signalAspects```
- ```conditions``` is a dictionary of conditions where:
  - Keys are ```condition codes```
  - Values are lists of parameters depending on condition codes. If condition code does not require parameters, it should be an empty list.

The order of the list of states is very important since every state will be evaluated in the order of the list to define the active aspect of the signal. The first state which matches all conditions defined in it is selected as the active aspect of the signal.

Available ```condition codes``` are:
- NEXT_ROUTE_ACTIVE (no parameters): A route must be set from this signal
- PREVIOUS_ROUTE_ACTIVE (no parameters): A route must set to this signal
- TRAIN_NOT_PRESENT_ON_NEXT_ROUTE (no parameters): No trains must be present on the route starting from this signal or between this signal and the next signal on the line if no route is set.
- TRAIN_NOT_PRESENT_ON_ITEMS (no parameters in tsl file): No train must be present on any items defined in "No train params" property (in TS2 editor).
- TRAIN_PRESENT_ON_ITEMS (no parameters in tsl file): Trains must be present simultaneously on all items defined in "Train present params" property (in TS2 editor).
- ROUTES_SET (no parameters in tsl file): At least one route defined in "route set params" property in TS2 editor must be set.
- NEXT_SIGNAL_ASPECTS (parameters = list of signal aspect names): The signal at the end of the route starting at this signal (or the next signal on the line if no route is set) must show one of the aspect given in the list.

### What can we do with it?

This system is very flexible to create all sorts of signals even with complicated logic. I've created the following new signals in just a few minutes (see UK.tsl and FR.tsl):
- Automatic signals : no need to setup a route from these signals, they just follow the next signal and check if there a train in the block
- UK "approach release from red" signals : I like this one because it seemed so difficult to setup with the previous signal system and it is now so easy to do.
- French BAL signals : I made them use almost all 6 "light bulbs" in turn, so it's interesting to see how signal aspects work.

### What's next?

- [ ] Allow routes to be set between two signals that are not contiguous which will allow to make intermediate shunting signals and also reminder signals.
- [ ] Implement 'running on sight' for trains to enable train shunting and permissive red signals.

**Comment by pedromorgan** [comment]
WOW// Looking hard...

Gets beer outta fridge... after a quick smoke.. ;-)))))))))))))))))))

**Comment by pedromorgan** [comment]
Ok Can I ask a muppet question..

1) which part of the above is the "way it works in real life"..

??
And which parts are we using for faking ??

**Comment by npiganeau** [comment]
All I can say is that I only know about real signalling what I read on the Internet (such as the links you put for newbies on the homepage), and I have no idea how this is done internally in real computer-controlled signalling centres (except that they have all sorts of interlocking systems that I don't have (yet) in TS2).

So from my reading, I get to know how different kinds of signals behave in real life. Reading SimSig's doc on their simulations is also quite interesting because it gives you ideas of peculiar signals that can exist in UK (e.g. in real liverpool street there is an axle counter that would allow berthing on patforms 16 to 18 only if your train is 8 coaches or less...).

Then seeing that there were so many different kinds of signals, I wanted to make a generic way of describing complex signals rather than hardcoding them one by one. That's were I got the idea of splitting signal aspects from signal types and to have a 'condition stack' to compute which aspect should be shown. So its my own invention from A to Z and I don't think it is any close to what's in real life, but who knows ?

**Comment by Sharpe49** [comment]
Well, technically, I'm a French train signalling engineer... so if you have any question, don't hesitate !
:)

**Comment by pedromorgan** [comment]
If I wanted to pitch to reate a multiline system as an "education project".. includind bring your honby trainsets et all..

How do we make the inerlocking work is my curious..

**Comment by pedromorgan** [comment]
Is ther

**Comment by npiganeau** [comment]
Actually, this is not completely true that there is no interlocking in TS2: there is a simple "route interlocking".

If we simplify, that's the way it works in TS2:
- Before you can set a route, the system looks if the route can be activated, i.e. that there is no conflicting route already set up. I suppose that is also what is happening in real life more or less.
- Signals have there "own lives" they check which aspect they should have each time they are triggered to do so, and they do it as explained above, taking into account the different routes set and the positions of the different trains. I suppose that this is not the real life system since this is not fool proof at all (!).

One of the main difference between TS2 and real life is that in real life each area is coded according to its real layout whereas in TS2 on the software side, I have no idea what the layout looks like (because it is made by the user in the simulation file). So in real life interlocks are hard-coded or hard-wired (or I hope they are, otherwise I won't jump on a train any more :-) ) whereas in TS2, interlocks are calculated dynamically by "walking along" the route you want to set and check if another route is not crossing it. You can see the code of this "dynamic interlocking" [there](https://github.com/ts2/ts2/blob/master/ts2/routing/route.py#L238).

Interlocks that are not implemented (yet?) in TS2 are:
- Points interlocks: points are only changed by route setting in TS2 and only the points necessary for this route are changed, not the points of the opposite line. Which means a train can cross an already set route if they SPAD.
- Approach locking: You can release a route even if the train is in front of the signal, setting the signal to danger and setup another (conflicting) route immediately. That will make the train SPAD and run into the conflicting route.
- Overlaps: In the UK, routes lock a portion of track (including points) ahead of a danger signal in case the train SPAD.

@Sharpe49 What are actually the different interlocking/safety systems in real life ? Are there overlaps in France or anything equivalent ?

**Comment by npiganeau** [comment]
@Sharpe49 
I have two other questions for you regarding French signalling:

**1. Train driver behaviour**
How are train drivers supposed to behave when facing a yellow warning aspect? On the internet it says: "He should to do what is necessary to be able to stop before the next signal". 
What I have coded is that trains keep their current speed as much as possible and reduce their speed constantly at x m/s² defined by the standard braking of the train type so as to stop exactly just a few meters from the next signal. But I could have made them reduce speed ASAP to let's say 30 km/h and keep this speed until the next signal.

What is the more realistic behaviour ?

**2. French sidings signals**
I've seen that siding signals have basically two aspects : white proceed, purple stop. 

But how is signalled the entrance of a siding from the main line ? Does the signal show white ? Then what would be the previous signal on the main line ?

And when exiting the sidings onto the main line ? Does the signal show white or green ?

**Comment by Sharpe49** [comment]
I will first answer about the question in the latest message :

###### 1. Train driver behaviour
Here's the document about the [VISA](http://www.securite-ferroviaire.fr/sites/default/files/users/reglementations/pdf/macacab1an1v1signe.pdf) which is what the driver is supposed to do when crossing a signal with a warning aspect.
He has to slow down to 30 km/h and, if the signal has opened, maintain this speed until he has crossed the signal.

###### 2. French sidings signals
At the entrance of a siding, you can have : 
- RR+A and G sign (Garage) or D sign (Dépôt) : the G/D sign tells the driver he has to shunt
- A and G/D sign if the current speed limit is 30 km/h
- S then M (approach control) : the train has to activate a detector for the M aspect to be presented

At the exit, it depends if the next signal is near or not.
If it is near, it is possible to leave the siding with the M aspect.
Otherwise, the presented aspect is the block aspect (S A VL).

**Comment by Sharpe49** [comment]
Firstly, I will talk about the steps that are necessary to open a signal.
1) Preparation : Points are moved if they are not locked.
2) Formation : If all of the points can be commanded to move in the right position, the route is formed.
3) Locking : The points are locked once the route is formed. They are locked by what we call transits. The transits are shown on the track plan.
4) Establishment : The positions of the points are checked and other conditions are checked (such as authorizations, some track circuits, direction locking, protection areas, etc.). Once the route is established, the signal opens.

A point is also locked if the train is on the track circuit of the point.

######Approach locking 
There are several types of "approach" locking : 
- The approach locking (EAp : Enclenchement d'Approche) is the most common type. If the train isn't in the approach area (ZAp), the route can be destroyed immediately. If the train is in the approach area, the route is locked. If you want to destroy it, you have to close the signal (using a FC (Fermeture Carré) switch), try to destroy the route once. A timer is launched (1 or 3 minutes depending on the speed). Once the timer is finished, you can retry to destroy the route. If the signal has not been opened, the destruction is done immediately.
- The route locking (EPa : Enclenchement de Parcours) is used when there is no approach area. If the signal has not been opened, the destruction is done immediately. Otherwise, the route is locked. In this case, you have to do the same procedure as in the approach locking in order to destroy the route.
- Finally, there is the timed manual destruction (DMT : Destruction Manuelle Temporisée). If the track circuit before the signal is clear, the destruction is done immediately. Otherwise, you have to do the same procedure.

######Overlaps
In France there are overlaps on the Paris metro signalling and in the TVM system on high speed lines. They are not used on classic lines.

**Comment by npiganeau** [comment]
Thanks @Sharpe49 for the infos !

So I changed the train behaviour in front of an A to reduce speed to 30 km/h ASAP, that's the closest I can do with the current implementation. Also, when the "driver" will "see" the next signal (that is ```defaultSignalVisibility``` metres before the signal, "he" will accelerate again if the signal has cleared in between, so he will behave as if there were continuous speed control.

I've also added FR siding signals:
- FR_SIDING : A standard CV / M signal. Shows M if a route is set from it, CV otherwise. Does not check if train is present. Can be used as an siding exit signal if to leave on M.
- FR_SIDING_OUT : For siding exit signal that will show standard block aspects if a route is set, and CV otherwise.

I also modified FR_BAL_RR30 so that it can be used as a siding entrance signal:
- Either showing A+RR (sorry, no way to show G/D...)
- Or S then M approach control.

I'll have to find a way to do A + G/D, but is not straightforward with the signal implementation.

You can see all of them in action in signal-test.ts2 with the entrance signal as S then M.

**Comment by Sharpe49** [comment]
There's something weird if I put several RR signals one after another.
The announcing signal should only present RR+A if a diverging route is set and the next signal is RR. Otherwise, the signal should display R.

Also, there can be Cv M signals on the main tracks. They are used for trains that have to move backwards on a different route after having crossed the signal. Their aspect is ignored for the calculation of the aspect of the main signals but if a route is set across a maneuver signal, they have to open too.

Finally, it must be possible to create route with no signal at its extremity. The EndItem must be accepted as a destination.

**Comment by npiganeau** [comment]
- First point, you're right. See my comment on your PR #41 
- For Cv/M on main track, ok I'll create a new signal type. I'll probably need to create a new condition too (ROUTE_SET_ACROSS). BTW, how do they open if next signal main line signal is A, S or C?
- For creating a route to enditem, the current logic is to use a *_TP signal as the last signal (see signal-test). Indeed, if we created a route to the enditem, we could not know which aspect to show on the previous signal.

**Comment by pedromorgan** [comment]
Guys.. your ahead of me..

I kida want to create a "guide to signalling" starter..

Kinda want to extract the above.. "how it works stuff".. ...

As an education et all.. 

Is that ok.. ??

**Comment by npiganeau** [comment]
@pedromorgan Ok. No problem !

**Comment by npiganeau** [comment]
@Sharpe49 I've created a new condition called ROUTE_SET_ACROSS. It is true if a route is defined across this signal, in the same direction but neither starting nor ending at this signal. I've updated my post above with this info.

I'll just let you create the signal Cv/M signal on line you described above, because it is not very clear yet in my mind.

**Comment by Sharpe49** [comment]
Is it normal that, if I put an automatic signal before a TP signal, the TP signal never opens ?

**Comment by Sharpe49** [comment]
And for the Cv M signal, it's what we call a intermediary signal.
Example : 
We have 3 signals called A, B and C.
A and C are C S A VL signals.
B is a Cv M signal.
There are 2 routes : from A to C and from B to C.
If a train is between A and B. You have to set the B-C route. Signal B is at the M aspect.
The train is shunting (mandatory).

If a train is in front of signal A, you can set the A-C route. If the C signal is closed, A is at the A aspect. B is at the M aspect.
If the train is shunting, the driver will look at the aspect of the Cv M signal. Otherwise, he will ignore it.

**Comment by npiganeau** [comment]
Just fixed the bug for TP not opening when after an automatic signal (it needed a trigger...)

**Comment by npiganeau** [comment]
And now added a new condition to FR_SIDING when a route is set over it on a main track. 

Check signal-test to tell me if this is the expected behaviour (the Cv/M before the TP signal)

**Comment by Sharpe49** [comment]
Yes, it is the expected behaviour.
There's something weird about the train though. It is slowing down when it is approaching the intermediary signal.

**Comment by Sharpe49** [comment]
Weird, when I load the Gretz simulation with the TSL file I have put in my pull request, this error is shown : 
Traceback (most recent call last):
  File "C:\Users\Cedric\Desktop\ts2\ts2\mainwindow.py", line 240, in loadSimulation
    self.simulation = simulation.load(self, file)
  File "C:\Users\Cedric\Desktop\ts2\ts2\simulation.py", line 115, in load
    simulation.initialize(simulationWindow)
  File "C:\Users\Cedric\Desktop\ts2\ts2\simulation.py", line 171, in initialize
    route.initialize(self)
  File "C:\Users\Cedric\Desktop\ts2\ts2\routing\route.py", line 133, in initialize
    self.activate(True)
  File "C:\Users\Cedric\Desktop\ts2\ts2\routing\route.py", line 262, in activate
    self.beginSignal.nextActiveRoute = self
  File "C:\Users\Cedric\Desktop\ts2\ts2\scenery\signals\signalitem.py", line 356, in _setNextActiveRoute
    self.updateSignalState()
  File "C:\Users\Cedric\Desktop\ts2\ts2\scenery\signals\signalitem.py", line 551, in updateSignalState
    self._activeAspect = self.signalType.getAspect(self)
  File "C:\Users\Cedric\Desktop\ts2\ts2\scenery\signals\signalitem.py", line 794, in getAspect
    if state.conditionsMet(signalItem, params):
  File "C:\Users\Cedric\Desktop\ts2\ts2\scenery\signals\signalitem.py", line 732, in conditionsMet
    if not applicableSolvers[conditionName](signalItem, parameters):
  File "C:\Users\Cedric\Desktop\ts2\ts2\scenery\signals\signalitem.py", line 1073, in solver
    return any([(r.getRouteState() != 0) for r in routes])
  File "C:\Users\Cedric\Desktop\ts2\ts2\scenery\signals\signalitem.py", line 1073, in <listcomp>
    return any([(r.getRouteState() != 0) for r in routes])
  File "C:\Users\Cedric\Desktop\ts2\ts2\routing\route.py", line 197, in getRouteState
    if self.beginSignal.nextActiveRoute is not None and \
  File "C:\Users\Cedric\Desktop\ts2\ts2\routing\route.py", line 169, in beginSignal
    return self._positions[0].trackItem
IndexError: list index out of range

**Comment by Sharpe49** [comment]
I can't understand what's happening.

**Comment by npiganeau** [comment]
Check your routes, maybe one is badly defined. Or I have a bug. I'll check.

**Comment by npiganeau** [comment]
I've just fixed the train slowing down when seeing a M aspect on main line: I created a new FR_M_DISCARD aspect that shows up like a M but is considered VL by the train.

**Comment by npiganeau** [comment]
Fixed indexError: list index out of range. It was a bug on my side that appeared after creating ROUTE_SET_ACROSS condition

**Comment by Sharpe49** [comment]
After having done some modifications, I get another error. I think it comes from my data.
Is it possible to have some checks during the save operation ?

Here's the error : 
Traceback (most recent call last):
  File "C:\Users\Cedric\Desktop\ts2\ts2\mainwindow.py", line 240, in loadSimulation
    self.simulation = simulation.load(self, file)
  File "C:\Users\Cedric\Desktop\ts2\ts2\simulation.py", line 115, in load
    simulation.initialize(simulationWindow)
  File "C:\Users\Cedric\Desktop\ts2\ts2\simulation.py", line 177, in initialize
    ti.setupTriggers()
  File "C:\Users\Cedric\Desktop\ts2\ts2\scenery\signals\signalitem.py", line 564, in setupTriggers
    trigger(self)
  File "C:\Users\Cedric\Desktop\ts2\ts2\scenery\signals\signalitem.py", line 1092, in trigger
    signalItem.simulation.routes[tiId].routeSelected.connect(
KeyError: 30

**Comment by npiganeau** [comment]
@Sharpe49 You're still one bug ahead of me ! 

I've just fixed the bug which made A+RR30 show instead of R30 in gretz simulation signal 3609. It was a pb in my code, the signal definition in tsl file was OK.

**Comment by npiganeau** [comment]
For the KeyError: 30, it probably is because you have one signal that checks when route 30 is set to display an aspect, and you deleted route 30 afterwards.

I'll see what I can do to have some checks when saving.

**Comment by Sharpe49** [comment]
Ok, so KeyError is the ID of the object that is searched...
I will try to find which signal provokes this error.

**Comment by Sharpe49** [comment]
Alright, I corrected the signal and tested your modification. Works perfectly.
I will try to add more trains now.
I think the next step for Gretz... will be the blinking aspects.

**Comment by npiganeau** [comment]
News from last devs of this week-end

- Trains now know how to run on sight. If you have 2 trains in the same block they will not run one on top of the other as before ! I'm just a few lines of code from a working permissive block system (such as french BAL), that is a block system where the driver is allowed to cross a danger signal after having stopped if he runs on sight with an appropriate speed limit.

- Signals change aspect to danger as soon as the train head crosses them like in real life (before, in ts2, we had to wait for the train tail to cross the signal for this to occur)

- Trains keep the speed limit given by a signal after crossing it, until they see the next signal (which was not the case before)

**Comment by npiganeau** [comment]
Done it! We have working permissive blocks now and they have been implemented with FR_BAL signals. When a train faces a S (single red) signal aspect, it stops in front of it, then 30 seconds later, it starts again with a limited speed of 30 km/h. All this behaviour is 100% configurable in TSL files.

Now, I will stop working on signals for a while to concentrate on the other topics of ts2. 

This is the todo list yet to be done for signals :

- [ ] Add shunt signals and banner repeaters to UK tsl files
- [ ] Update liverpool street simulation to include shunt signals, repeaters, call on routes, etc.

**Comment by npiganeau** [comment]
@Sharpe49 I've added checks when saving a sim in the editor (or when loading in game) to see if the sim is valid, especially for route and track item reference in signals.
